af/fastpath/fastpath/normalize.py

```diff
@@ -0,0 +1,811 @@
+        entry["input"] = hashed_fingerprint
+        return entry
+    return entry
+
+
+def normalize_entry(entry, bucket_date, perma_fname, esha):
+    """Autoclaving
+    """
+    hashuuid = esha[:16]  # sha1 is 20 bytes
+
+    if isinstance(entry.get("report"), dict):
+        entry.update(entry.pop("report"))
+
+    test_name = entry.get("test_name", "invalid")
+    test_name = test_name_mappings.get(test_name, test_name.lower())
+    entry["test_name"] = test_name
+
+    entry["bucket_date"] = bucket_date
+
+    if not entry.get("id"):
+        entry["id"] = str(uuid.UUID(bytes=hashuuid))
+    entry["report_filename"] = perma_fname
+
+    # Ensure all the keys in the schema are present
+    for key in schema:
+        entry[key] = entry.get(key, None)
+
+    if entry.get("data_format_version", "0.1.0") == "0.2.0":
+        if entry["test_keys"] is None:
+            entry = nest_test_keys(entry)
+        return entry
+
+    test_start_time = datetime.fromtimestamp(entry.pop("start_time", 0))
+    try:
+        tst = entry.pop("test_start_time")
+        # This is the old test_start_time key that now is called
+        # "measurement_start_time"
+        if isinstance(tst, float):
+            measurement_start_time = datetime.fromtimestamp(tst)
+        elif tst is None:
+            measurement_start_time = test_start_time
+        else:
+            test_start_time = datetime.strptime(tst, "%Y-%m-%d %H:%M:%S")
+            measurement_start_time = datetime.strptime(
+                entry.get("measurement_start_time"), "%Y-%m-%d %H:%M:%S"
+            )
+    except KeyError:
+        # Failback to using the start_time
+        measurement_start_time = test_start_time
+
+    entry["measurement_start_time"] = measurement_start_time.strftime(
+        "%Y-%m-%d %H:%M:%S"
+    )
+    entry["test_start_time"] = test_start_time.strftime("%Y-%m-%d %H:%M:%S")
+
+    entry["data_format_version"] = "0.2.0"
+
+    if isinstance(entry.get("options", []), dict):
+        entry["options"] = entry["options"].get("subargs", [])
+
+    entry = nest_test_keys(entry)
+
+    assert test_name not in test_categories["process"]
+
+    # Some values overlap across categories so multiple ifs are needed
+    if test_name in test_categories["httpt"]:
+        entry = normalize_httpt(entry)
+    if test_name in test_categories["dnst"]:
+        entry = normalize_dnst(entry)
+    if test_name in test_categories["scapyt"]:
+        entry = normalize_scapyt(entry)
+    if test_name == "captive_portal":
+        entry = normalize_captive_portal(entry)
+    if test_name == "tls_handshake":
+        entry = normalize_tls_handshake(entry)
+
+    # TODO: tests these in
+    # test_normalize_yaml_sanitise_tcp_connect_bridge_reach
+    if entry["test_name"] == "tcp_connect":
+        entry = sanitise_tcp_connect(entry, BRIDGE_DB)
+
+    elif entry["test_name"] == "bridge_reachability":
+        entry = sanitise_bridge_reachability(entry, BRIDGE_DB)
+
+    return entry
+
```

Comment:
It looks like a lot of the code in here comes from `daily_workflow.py`. I think we should find a way to avoid having this code be duplicated. Ideally both the fastpath and the batch pipeline would share the same module for doing this type of normalisation or at least some part of this code should be shared by both components.

Reply:
In the long term, yes. I plan to have this code well tested, merged and run and then backport it.